msal/src/main/java/com/microsoft/identity/client/Account.java

```diff
@@ -22,73 +22,102 @@
+    private String mHomeOid;
+    private String mHomeTenantId;
     private String mEnvironment;
 
-    Account() {
-        // Empty constructor
+    public Account(@Nullable final String rawIdToken) {
+        mRawIdToken = rawIdToken;
     }
 
-    /**
-     * Sets the account id.
-     *
-     * @param accountId The IAccountIdentifier to set.
-     */
-    void setAccountIdentifier(final IAccountIdentifier accountId) {
-        mAccountIdentifier = accountId;
+    void setId(@Nullable final String id) {
+        mHomeOid = id;
     }
 
+    @NonNull
     @Override
-    public IAccountIdentifier getAccountIdentifier() {
-        return mAccountIdentifier;
-    }
+    public String getId() {
+        String id;
+
+        if (null != mRawIdToken && null != getClaims()) {
+            id = (String) getClaims().get(MicrosoftIdToken.OBJECT_ID);
+        } else {
+            id = mHomeOid;
+        }
 
-    /**
-     * Sets the home account id.
-     *
-     * @param homeAccountId The IAccountIdentifier to set.
-     */
-    void setHomeAccountIdentifier(final IAccountIdentifier homeAccountId) {
-        mHomeAccountIdentifier = homeAccountId;
+        return id;
     }
 
-    @Override
-    public IAccountIdentifier getHomeAccountIdentifier() {
-        return mHomeAccountIdentifier;
+    void setTenantId(@Nullable final String tenantId) {
+        mHomeTenantId = tenantId;
     }
 
-    /**
-     * Sets the username.
-     *
-     * @param username The username to set.
-     */
-    void setUsername(final String username) {
-        mUsername = username;
+    @Nullable
+    String getTenantId() {
+        return mHomeTenantId;
     }
 
-    @Override
-    public String getUsername() {
-        return mUsername;
+    @Nullable
+    String getHomeAccountId() {
+        return getId() + "." + mHomeTenantId;
     }
 
-    /**
-     * Sets the environment.
-     *
-     * @param environment The environment to set.
-     */
-    void setEnvironment(final String environment) {
+    void setEnvironment(@NonNull final String environment) {
         mEnvironment = environment;
     }
 
-    @Override
-    public String getEnvironment() {
+    @Nullable
+    String getEnvironment() {
         return mEnvironment;
     }
+
+    @Nullable
+    @Override
+    public Map<String, ?> getClaims() {
+        Map<String, ?> claims = null;
+
+        if (null != mRawIdToken) {
+            claims = getClaims(mRawIdToken);
+        }
+
+        return claims;
+    }
+
+    private static Map<String, ?> getClaims(@NonNull final String rawIdToken) {
+        final String methodName = ":getClaims(String)";
+
+        final Map<String, Object> result = new HashMap<>();
+
+        try {
+            final JWT jwt = JWTParser.parse(rawIdToken);
+            final JWTClaimsSet claimsSet = jwt.getJWTClaimsSet();
+            result.putAll(claimsSet.getClaims());
+        } catch (ParseException e) {
+            Logger.error(
+                    TAG + methodName,
```

Comment:
Instead of explicitly capturing the methodName wouldn't be better to use something like getEnclosingMethod()

Reply:
I believe the reason why this isn't being done is because that method requires the use of reflection.  Also which object did you want to get the class for within the method to determine the enclosing method?

Reply:
In general reflection comes at fairly high price in terms of performance and the tradeoff here is this:

```java
String methodName = ":getFoo";
```

vs

```java
String methodName = getClass().getEnclosingMethod().getName();
```

IntelliJ / AndroidStudio is smart enough to change these values when autorefactor is used to renamed methods. Since string allocations are pretty cheap in terms of perf I'm inclined to leave as-is